Update `presubmit` workflow to `gradle-build-action@v2`

bigdaz opened the pull request:
## Proposed Changes

  - Replace use of `gradle-command-action@v1` with `gradle-build-action@v2`.
  - Only write to the cache from the `androidx-main` branch. See https://github.com/gradle/gradle-build-action#only-write-to-the-cache-from-the-default-branch.
  - Enable build-cache and build-scans for `lint` workflow job
  - Disable Gradle configuration-cache due to errors reusing persisted state.
  - Remove use of `findAffectedModules` for build avoidance in `build-modules` job. This was causing failures to be hidden in subsequent workflow runs.
  - Add a single link to all build scans for a workflow run.

## Testing

Test: via GitHub Actions run

bigdaz committed:
Update `presubmit` workflow to `gradle-build-action@v2`

bigdaz committed:
Don't save Gradle User Home for branch workflows

GitHub cache entries are not shared between builds on different branches.
This means that identical cache entries will be stored separately for different branches.
The exception to the is cache entries for the default (androidx-main) branch can be read by
actions invoked for other branches.

To avoid having private branch-specific entries evict shared entries,
we configured branch workflows to use the cache read-only.

liutikas commented on .github/workflows/presubmit.yml:
```diff
@@ -3,6 +3,9 @@ on:
   pull_request:
 
+env:
+  GRADLE_BUILD_ACTION_CACHE_DEBUG_ENABLED: true
```
Please add a comment on what this does with a link to docs https://github.com/marketplace/actions/gradle-build-action#cache-debugging-and-analysis

bigdaz replied:
Done

liutikas commented on .github/workflows/presubmit.yml:
```diff
@@ -37,13 +41,18 @@ jobs:
           fi
           echo "::set-output name=repository::$REPOSITORY"
+      - name: Publish build scans link
+        run: echo "::notice title=All build scans for workflow::https://ge.androidx.dev/scans?search.names=CI%20run&search.values=$GITHUB_RUN_ID"
```
Oh this is nice!

liutikas commented on .github/workflows/presubmit.yml:
```diff
@@ -94,35 +103,26 @@ jobs:
+          gradle-home-cache-strict-match: true
+          # Limit the size of the cache entry
+          gradle-home-cache-excludes: | 
+            caches/jars-9
```
Please add comments on what is in these directories so we know what we are dropping. Ideally link to Gradle docs.

bigdaz replied:
Done. Unfortunately these directories are undocumented.

liutikas commented on .github/workflows/presubmit.yml:
```diff
@@ -19,6 +22,7 @@ jobs:
             -Dorg.gradle.internal.http.socketTimeout=60000                  \
             -Dorg.gradle.internal.repository.max.retries=20                 \
             -Dorg.gradle.internal.repository.initial.backoff=500            \
+            --no-configuration-cache                                        \
```
Are we disabling configuration to speed up CI?

bigdaz replied:
There are 2 reasons this needs to be disabled:
1. Currently, the `buildOnServer` task fails when restored from the configuration cache. This happens when I run the build locally: https://ge.androidx.dev/s/zist7pgwj3qfy/console-log#L303
2. The configuration-cache state contains links directly into the `caches/jars-9` and `caches/transforms-3` directories that are currently being excluded from the Gradle User Home save/restore.

We can fix the first issue, but until we are sure we can remove the cache excludes we cannot use configuration caching. I'll add a comment to this effect.

Note that there isn't yet a way to tell `gradle-build-action` to avoid save/restore of the configuration-cache state: if the configuration-cache is written in a build invocation, the action will save it and restore on the next run. So `--no-configuration-cache` is the best way to avoid this problem.

bigdaz replied:
Comment added to workflow file.

liutikas commented:
You will need to rebase this PR since I removed AffectedModuleDetector.

bigdaz committed:
Add explanatory comments to workflow

bigdaz committed:
Omit build-scan link for PR from forked repo

For a PR from a forked repo, the build-scan publish secret won't be set,
and no build scans will be published. So don't emit a link to all build
scans in this case.

yigit commented on .github/workflows/presubmit.yml:
```diff
@@ -171,42 +179,33 @@ jobs:
+          # Limit the size of the cache entry.
+          # These directories contain instrumented/transformed dependency jars which can be reconstructed relatively quickly.
+          gradle-home-cache-excludes: | 
+            caches/jars-9
```
Do we need something like jars-* instead of 9? (Idk if * is supported but 9 seems very random and could break).

bigdaz replied:
Unfortunately this parameter doesn't (yet) support wildcards. I'll add an issue to `gradle-build-action` for this.

The `jars-9` is tied to a set of Gradle versions, and will only be incremented when the instrumentation algorithm changes in Gradle. If you update to a new Gradle version with a new `jars-` version, then the cache entry will start to contain these files. Things will keep working, but we might start to see cache entry eviction based on hitting the 10Gb limit.

In the end, I'm not certain that these exclusions are necessary, but in my testing they allowed a bit of headroom for growth in entry size without leading to premature eviction. I'd like to get this in and then start to monitor the real-world behaviour, tweaking as necessary.

bigdaz replied:
The `gradle-build-action` makes it easy to monitor the size of entries being uploaded: eg https://github.com/bigdaz/androidx/runs/4974841719?check_suite_focus=true#step:17:221. So this will allow us to monitor how well the Gradle User Home caching is performing.
I'm sure further tweaks will be required.

yigit commented:
Overall LGTM, just not sure about the numbers in directories
